cmd/promtool/main.go

```diff
@@ -255,57 +257,40 @@ func checkConfig(filename string) ([]string, error) {
 	}
 
 	for _, scfg := range cfg.ScrapeConfigs {
-		if err := checkFileExists(scfg.HTTPClientConfig.BearerTokenFile); err != nil {
-			return nil, errors.Wrapf(err, "error checking bearer token file %q", scfg.HTTPClientConfig.BearerTokenFile)
-		}
-
-		if err := checkTLSConfig(scfg.HTTPClientConfig.TLSConfig); err != nil {
+		if err := config.ValidateHTTPClientConfig(&scfg.HTTPClientConfig); err != nil {
 			return nil, err
 		}
 
-		for _, kd := range scfg.ServiceDiscoveryConfig.KubernetesSDConfigs {
-			if err := checkTLSConfig(kd.HTTPClientConfig.TLSConfig); err != nil {
-				return nil, err
-			}
-		}
-
-		for _, filesd := range scfg.ServiceDiscoveryConfig.FileSDConfigs {
-			for _, file := range filesd.Files {
-				files, err := filepath.Glob(file)
-				if err != nil {
-					return nil, err
+		for _, c := range scfg.ServiceDiscoveryConfigs {
+			switch v := c.(type) {
+			case interface{ Validate() error }:
```

Comment:
This is unnecessary complexity, validation should be happening at unmarshall time and we should also be avoiding optional interfaces.

Reply:
> This is unnecessary complexity, validation should be happening at unmarshall time

This is checking for the existence of files and validity of their content. Unmarshaling YAML shouldn't have the side-effect of reading TLS certificates from disk. What if I'm just shuffling YAML configs around on a server that doesn't have the TLS certificates (e.g. `prometheus-operator`)? Now I can't unmarshal YAML? Implementing this would be extremely complex given that the path resolution is unknown to the unmarshalers, and even if we managed it, it would be in direct contradiction to your other requirement of delaying path resolution until Discoverer construction.

> we should also be avoiding optional interfaces

It's optional because you objected to it being required.

It exists in the first place because a generic method of validating configs in `promtool` was requested during this review. Previously, only a subset of files in kubernetes configs were checked. Now all files in all configs are. I think this is a useful improvement to `promtool`, but again it's not a feature that I'm trying to champion.

Reply:
> Now I can't unmarshal YAML?

Yes, that's the existing behaviour. All this code is Prometheus-internal, and never used by anyone else.

> It's optional because you objected to it being required.

I object to the way it is implemented. Making it optional doesn't fix that, it only makes it more complicated.

Reply:
The existing behavior is _not_ to fail unmarshaling if the object contains a reference to a file that doesn't exist. The existing unmarshaling doesn't have the context necessary to resolve the file paths.

I reverted the previously requested changes to `promtool` and removed the `Validate() error` method. Now this whole thing is a behavioral no-op again and only addresses the issue it is originally intended to: dependencies.

If you want the successful conversion of a byte buffer into a go struct to be dependent on the existence and validity of arbitrary files on disk, I suggest you create a separate issue.

Reply:
`If you want the successful conversion of a byte buffer into a go struct to be dependent on the existence and validity of arbitrary files on disk, I suggest you create a separate issue.` @brian-brazil this is something I still plan to do something about, related to the other PR I opened. If this is a blocker for @abursavich's PR I can prioritize making that change sooner rather than later.

Reply:
> The existing behavior is not to fail unmarshaling if the object contains a reference to a file that doesn't exist. 

What matters is if the config works or not, what exact stage a broken config is detected at internally isn't important from the user standpoint. A failed config load would be better than a failed scrape of course.

> If this is a blocker for @abursavich's PR I can prioritize making that change sooner rather than later.

It's not per-se, but the current way we do things is a bit all over the place and both this PR's proposed changes and @cstyan's would aggravate the situation. I'm hoping to at least not make things worse in this PR, and best case have this in a much cleaner situation for @cstyan's work plus #5222 which @roidelapluie mentioned. Presuming we decide to implement #5222 - I'm aware of it but it is non-trivial and poses maintainability issues given the current code layout, issues which this PR could and possibly has to solve.

As it stands it looks like we're adding complexity and further layering violations to paper over a suboptimal design that has built up over time, rather than addressing the core design issues here.

Reply:
> What matters is if the config works or not, what exact stage a broken config is detected at internally isn't important from the user standpoint. A failed config load would be better than a failed scrape of course.

Here's a sketch of formalizing the Validater interface, implementing it on all of the config types recursively, and using it to fail fast: https://github.com/abursavich/prometheus/compare/config...abursavich:validate

Reply:
That seems like an okay approach, though several of the checks there should be caught at unmarhsall time, and that code should be down in common so it's properly recursive.